validator/index.js

```diff
@@ -25,50 +25,233 @@ const path = require('path');
- * @param {!string} fileOrUrl
- * @param {!function(string)} onSuccess
- * @param {!function(string)} onFailure
+ * Convenience function to detect whether an argument is a ULR or
+ * (perhaps) a local file.
+ * @param {!string} url
+ * @returns {!boolean}
  */
-function readFileOrDownload(fileOrUrl, onSuccess, onFailure) {
-  const digestChunksFrom = function(response) {
+function isHttpOrHttpsUrl(url) {
+  return url.startsWith('http://') || url.startsWith('https://');
+}
+
+/**
+ * Creates a promise which reads from a file.
+ * @param {!string} name
+ * @returns {!Promise<(!string|!Error)>}
+ */
+function readFromFile(name) {
+  return new Promise(function(resolve, reject) {
+    fs.readFile(name, 'utf8', (err, data) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(data);
+      }
+    });
+  });
+}
+
+/**
+ * Creates a promise which reads from a stream.
+ * @param {!stream.Readable} readable
+ * @returns {!Promise<!string|Error>}
+ */
+function readFromReadable(readable) {
+  return new Promise(function(resolve, reject) {
     const chunks = [];
-    response.setEncoding('utf8');
-    response.on('data', function(chunk) { chunks.push(chunk); });
-    response.on('end', function() { onSuccess(chunks.join('')); });
-  };
-  if (fileOrUrl === '-') {
-    digestChunksFrom(process.stdin);
-    process.stdin.resume();
-    return;
-  }
-  if (fileOrUrl.startsWith('http://') || fileOrUrl.startsWith('https://')) {
-    const clientModule = fileOrUrl.startsWith('http://') ? http : https;
-    const req = clientModule.get(fileOrUrl, (response) => {
-      if (response.statusCode !== 200) {
-        onFailure('HTTP ' + response.statusCode);
-        return;
+    readable.setEncoding('utf8');
+    readable.on('data', (chunk) => { chunks.push(chunk); });
+    readable.on('end', () => { resolve(chunks.join('')); });
+    readable.on('error', (error) => { reject(error); });
+  });
+}
+
+/**
+ * Creates a promise which reads from standard input.
+ * @returns {!Promise<!string|Error>}
+ */
+function readFromStdin() {
+  return new Promise(function(resolve, reject) {
+    readFromReadable(process.stdin)
+        .then((data) => {
+          process.stdin.resume();
+          resolve(data);
+        })
+        .catch(reject);
+  });
+}
+
+/**
+ * Creates a promise which reads from a URL or more precisely, fetches
+ * the contents located at the URL by using the 'http' or 'https' module.
+ * Any HTTP status other than 200 is interpreted as an error.
+ * @param {!string} url
+ * @returns {!Promise<!string|Error>}
+ */
+function readFromUrl(url) {
+  return new Promise((resolve, reject) => {
+    const clientModule = url.startsWith('http://') ? http : https;
+    clientModule.get(url, (response) => {
+      if (response.statusCode != 200) {
+        response.resume();
+        reject(new Error('HTTP Status ' + response.statusCode));
+      } else {
+        readFromReadable(response).then(resolve);
```

Comment:
Putting readFromReadable into there is neat!
I've added a comment about .resume(). My reading of the documentation is that it's needed, both before and after the code change you suggested. The way I read the new way with the readFromReadable at the end is that it only triggers if resolve was called, not for reject, so since we did register a response callback we need to consume with data/end/error or call resume().